src/contracts/yas_pool.cairo

```diff
@@ -84,25 +185,304 @@ mod YASPool {
         /// @dev price is represented as a sqrt(amount_token_1/amount_token_0) Q64.96 value
         /// @param sqrt_price_X96 the initial sqrt price of the pool as a Q64.96
         fn initialize(ref self: ContractState, sqrt_price_X96: FixedType) {
-            // The initialize function should only be called once. To ensure this, 
+            // The initialize function should only be called once. To ensure this,
             // we verify that the price is not initialized.
             let mut slot_0 = self.slot_0.read();
             assert(slot_0.sqrt_price_X96.is_zero(), 'AI');
 
             slot_0.sqrt_price_X96 = sqrt_price_X96;
             slot_0.tick = TickMath::get_tick_at_sqrt_ratio(sqrt_price_X96);
             slot_0.fee_protocol = 0;
-            slot_0.unlocked = true;
             self.slot_0.write(slot_0);
 
+            self.unlocked.write(true);
+
             self.emit(Initialize { sqrt_price_X96, tick: slot_0.tick });
         }
+
+        fn swap(
+            ref self: ContractState,
+            recipient: ContractAddress,
+            zero_for_one: bool,
+            amount_specified: i256,
+            sqrt_price_limit_X96: FixedType,
+            data: Array<felt252>
+        ) -> (i256, i256) {
+            assert(amount_specified.is_non_zero(), 'AS');
+
+            let slot_0_start: Slot0 = self.slot_0.read();
+
+            let mut unlocked = self.unlocked.read();
+            assert(unlocked, 'LOK');
+            assert(
+                if zero_for_one {
+                    sqrt_price_limit_X96 < slot_0_start.sqrt_price_X96
+                        && sqrt_price_limit_X96 > TickMath::MIN_SQRT_RATIO()
+                } else {
+                    sqrt_price_limit_X96 > slot_0_start.sqrt_price_X96
+                        && sqrt_price_limit_X96 < TickMath::MAX_SQRT_RATIO()
+                },
+                'SPL'
+            );
+
+            unlocked = false;
+            self.unlocked.write(unlocked);
```

Comment:
Could we use a wrapper function called `lock` and `unlock` something in those lines? It would make everything cleaner.

Reply:
✅